File: go/fixture/go-mini/internal/handlers/heartbeat.go
package handlers

import (
	"encoding/json"
	"io"
	"net/http"

	"example.com/go-mini/internal/services"
)

// maxHeartbeatBytes bounds a heartbeat line; the longest legitimate line is
// well under a kilobyte.
const maxHeartbeatBytes = 4096

type heartbeatResponse struct {
	ID      string   `json:"id"`
	Score   int      `json:"score"`
	Changed bool     `json:"changed"`
	Tags    []string `json:"tags"`
}

// Heartbeat returns the heartbeat handler. The body is one raw heartbeat
// line, the tenant comes from the X-Tenant header and ?force=1 forces the
// reported status through.
func Heartbeat(svc *services.DeviceService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		raw, err := io.ReadAll(http.MaxBytesReader(w, r.Body, maxHeartbeatBytes))
		if err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": "unreadable body"})
			return
		}
		tenant := r.Header.Get("X-Tenant")
		if tenant == "" {
			tenant = "default"
		}
		force := r.URL.Query().Get("force") == "1"

		id, score, changed, tags, err := svc.ProcessHeartbeat(string(raw), tenant, force)
		if err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]string{"error": err.Error()})
			return
		}
		if tags == nil {
			tags = []string{}
		}
		writeJSON(w, http.StatusOK, heartbeatResponse{ID: id, Score: score, Changed: changed, Tags: tags})
	}
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

File: py/fixture/py-mini/internal/handlers/heartbeat.py
"""The heartbeat endpoint."""

from http import HTTPStatus

from internal.handlers.mux import HandlerFunc, Request, Response, write_json
from internal.services.device_service import DeviceService

# Bounds a heartbeat line; the longest legitimate line is well under a kilobyte.
_MAX_HEARTBEAT_BYTES = 4096

# The query value that forces the reported status through.
_FORCE = "1"


def heartbeat(svc: DeviceService) -> HandlerFunc:
    """Return the heartbeat handler.

    The body is one raw heartbeat line, the tenant comes from the X-Tenant
    header and ?force=1 forces the reported status through.
    """

    def handle(req: Request) -> Response:
        if len(req.body) > _MAX_HEARTBEAT_BYTES:
            return write_json(HTTPStatus.BAD_REQUEST, {"error": "unreadable body"})
        raw = req.body.decode("utf-8", errors="replace")
        tenant = req.headers.get("x-tenant") or "default"
        force = req.query.get("force") == _FORCE

        device_id, score, changed, tags, err = svc.process_heartbeat(raw, tenant, force)
        if err is not None:
            return write_json(HTTPStatus.BAD_REQUEST, {"error": err})
        if tags is None:
            tags = []
        return write_json(
            HTTPStatus.OK, {"id": device_id, "score": score, "changed": changed, "tags": tags}
        )

    return handle
